core/primitives/address_codec.cpp

```diff
@@ -0,0 +1,216 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "primitives/address_codec.hpp"
+
+#include <stdexcept>
+
+#include <cppcodec/base32_default_rfc4648.hpp>
+#include "codec/leb128/leb128.hpp"
+#include "common/visitor.hpp"
+#include "crypto/blake2/blake2b.h"
+
+namespace fc::primitives {
+
+  static const size_t kBlake2b160HashSize{20};
+  static const size_t kBLSHashSize{48};
+
+  using common::Blob;
+  using base32 = cppcodec::base32_rfc4648;
+
+  std::vector<uint8_t> encode(const Address &address) {
+    std::vector<uint8_t> res{};
+    res.push_back(address.network);
+    res.push_back(address.getProtocol());
+    std::vector<uint8_t> payload =
+        visit_in_place(address.data,
+                       [](uint64_t v) { return codec::leb128::encode(v); },
+                       [](const auto &v) {
+                         return std::vector<uint8_t>(v.begin(), v.end());
+                       });
+    res.insert(res.end(), payload.begin(), payload.end());
+    return res;
+  }
+
+  outcome::result<Address> decode(const std::vector<uint8_t> &v) {
+    if (v.size() < 3) return outcome::failure(AddressError::INVALID_PAYLOAD);
+    auto pos = v.begin();
+
+    auto net = static_cast<Network>(*(pos++));
+    if (!(net == Network::TESTNET || net == Network::MAINNET)) {
+      return outcome::failure(AddressError::UNKNOWN_NETWORK);
+    }
+
+    auto p = static_cast<Protocol>(*(pos++));
+    std::vector<uint8_t> payload(pos, v.end());
+    switch (p) {
+      case Protocol::ID: {
+        auto decoded = codec::leb128::decode<uint64_t>(payload);
+        if (!decoded) {
+          return outcome::failure(AddressError::INVALID_PAYLOAD);
+        }
+        return outcome::success(Address{net, decoded.value()});
+      }
+      case Protocol::SECP256K1: {
+        if (payload.size() != kBlake2b160HashSize) {
+          return outcome::failure(AddressError::INVALID_PAYLOAD);
+        }
+        Secp256k1PublicKeyHash hash{};
+        std::copy_n(payload.begin(), kBlake2b160HashSize, hash.begin());
```

Comment:
```suggestion
        std::copy_n(std::make_move_iterator(payload.begin()), kBlake2b160HashSize, hash.begin());
```
Move instead copy.
Here and further.

Reply:
* After pondering over. *
There is almost no sense to move basic types, but still move is always better.